types/time.go

```diff
@@ -2340,6 +2340,10 @@ func GetFormatType(format string) (isDuration, isDate bool) {
 
 func matchDateWithToken(t *MysqlTime, date string, token string, ctx map[string]int) (remain string, succ bool) {
 	if parse, ok := dateFormatParserTable[token]; ok {
+		if len(date) == 0 {
```

Comment:
https://github.com/pingcap/tidb/pull/12623/files#diff-3d97ef53b300d76ad11d0d73a7b7a7ebR2229

Is it better to move the change here?
If the next format token is not a valid one, what's MySQL's behavior?

For example:

```
select str_to_date('20190101', '%Y%m%d%!');
```

Reply:
```
mysql> select str_to_date('20190101', '%Y%m%d%!');
+-------------------------------------+
| str_to_date('20190101', '%Y%m%d%!') |
+-------------------------------------+
| 2019-01-01                          |
+-------------------------------------+
```